src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -700,11 +743,63 @@ private void EmitLanguageBlockNode(AssociativeNode node)
             throw new NotImplementedException();
         }
 
-        private void EmitRangeExprNode(AssociativeNode node)
+        private Type EmitRangeExprNode(AssociativeNode node)
         {
-            throw new NotImplementedException();
+            var range = node as RangeExprNode;
+            var fromNode = range.From;
+            var toNode = range.To;
+            var stepNode = range.Step;
+            var stepOp = range.StepOperator;
+            var hasAmountOperator = range.HasRangeAmountOperator;
+
+            //TODO If the from / to nodes are constants we can do some checks
+
+            //call generate range c# function.
+
+            IntNode op = null;
+            switch (stepOp)
+            {
+                case DSASM.RangeStepOperator.StepSize:
+                    op = new IntNode(0);
+                    break;
+                case DSASM.RangeStepOperator.Number:
+                    op = new IntNode(1);
+                    break;
+                case DSASM.RangeStepOperator.ApproximateSize:
+                    op = new IntNode(2);
+                    break;
+                default:
+                    op = new IntNode(-1);
+                    break;
+            }
+            var arguments = new List<AssociativeNode>
+            {
+                fromNode,
+                toNode,
+                stepNode ?? new NullNode(),
+                op,
+                AstFactory.BuildBooleanNode(stepNode != null),
+                AstFactory.BuildBooleanNode(hasAmountOperator),
+            };
+            //TODO shortest lacing?
+            //TODO without more refactoring we cannot reference the names of the generate range functions. fix that.
+            //maybe move to designscriptbuiltins and use type forwarding for enum?
+            var rangeExprFunc = AstFactory.BuildFunctionCall("GenerateRangeIL", arguments);
```

Comment:
yes, we discussed in the meeting and this was one of the options, I will send you the recording link - but this only works when the AST actually defines for us what the types are - in many cases the AST nodes will be identifiers with no type information.

Reply:
I've taken this approach, PTAL - I have yet to implement the fallback boxing method, but it will basically be the first version of this code that I wrote using objects.